src/library/scala/concurrent/Future.scala

```diff
@@ -413,8 +413,31 @@ trait Future[+T] extends Awaitable[T] {
    * @return        a `Future` with the result of the application of `f` to the results of `this` and `that`
    * @group Transformations
    */
-  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] =
-    flatMap(r1 => that.map(r2 => f(r1, r2)))(if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic)
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
+    val buffer = new AtomicReference[Either[T, U]]()
+    val outPromise = Promise[R]()
+    this.onComplete{
+      case scala.util.Success(v) =>
+        val isFirst = buffer.compareAndSet(null, Left(v))
+        if (!isFirst) {
+          val oldValue = buffer.get.getOrElse(throw new Exception("CANT HAPPEN"))
+          outPromise.complete(Try(f(v, oldValue)))
+        }
+
+      case scala.util.Failure(e) => outPromise.tryFailure(e)
+    }
+    that.onComplete{
+      case scala.util.Success(v) =>
+        val isFirst = buffer.compareAndSet(null, Right(v))
+        if (!isFirst) {
+          val oldValue = buffer.get.swap.getOrElse(throw new Exception("CANT HAPPEN"))
+          outPromise.complete(Try(f(oldValue, v)))
+        }
+
+      case scala.util.Failure(e) => outPromise.tryFailure(e)
+    }
+    outPromise.future
+  }
```

Comment:
@lihaoyi I think it could be possible to improve on the performance by avoiding wrapping, and avoiding compare-and-set.

```scala
  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      val buffer = new AtomicReference[Success[Any]]()
      val zipped = Promise[R]()
      this.onComplete {
        case left: Success[T] =>
          val right = buffer.getAndSet(left)
          if (right ne null)
            zipped.tryComplete(left.map(l => f(l, right.get.asInstanceOf[U])))
        case f: Failure[T] =>
          zipped.tryComplete(f.asInstanceOf[Failure[R]])
      }
      that.onComplete {
        case right: scala.util.Success[U] =>
          val left = buffer.getAndSet(right)
          if (left ne null)
            zipped.tryComplete(right.map(r => f(left.get.asInstanceOf[T], r)))
        case f: scala.util.Failure[U] =>
          zipped.tryComplete(f.asInstanceOf[Failure[R]])
      }
      zipped.future
    }
```

Reply:
@lihaoyi The solution above is correct, if you want to get rid of the allocation if the function(s) to `map` it is safe to do:

```scala
  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      val buffer = new AtomicReference[Success[Any]]()
      val zipped = Promise[R]()
      this.onComplete {
        case left: Success[T] =>
          val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
          if (right ne null)
            zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
        case f: Failure[T] =>
          zipped.tryComplete(f.asInstanceOf[Failure[R]])
      }
      that.onComplete {
        case right: Success[U] =>
          val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
          if (left ne null)
            zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
        case f: Failure[U] =>
          zipped.tryComplete(f.asInstanceOf[Failure[R]])
      }
      zipped.future
    }
```

Reply:
I think I ended up with more or less the same thing, except I used `complete` instead of `tryComplete` for the success-with-previous-value case case, since AFAICT that case can never happen when the `outPromise` has already been completed

Reply:
@lihaoyi 

this is de default implementation of `complete`: 

```scala
  def complete(result: Try[T]): this.type =
    if (tryComplete(result)) this else throw new IllegalStateException("Promise already completed.")
```

So it is faster to just call tryComplete directly since it will not fail.